Make parallel-netcdf an optional dependency, only compiled if MPAS is used

**Is your feature request related to a problem? Please describe.**

`parallel-netcdf` isn't needed unless the MPAS model is used. The interfaces to `parallel-netcdf` are already turned off in `esmf` and `netcdf-c`. But the default setting for `parallelio` is to enable the interface.

This package is causing increasing headaches, especially with the latest update to spack develop (#1607) on Cray systems. Also, it is one of the few packages that simply don't build if your shell isn't bash but ksh (yes, that case still exists).
```
1 error found in build log:
     19    checking for gawk... (cached) gawk
     20    checking for grep that handles long lines and -e... /usr/bin/grep
     21    checking for egrep... /usr/bin/grep -E
     22    checking for a sed that does not truncate output... /usr/bin/sed
     23    checking for sed handling option -i ... /usr/bin/sed -i''
     24    checking cc under /opt/cray/pe/mpich/8.1.26/ofi/intel/2022.1/bin... no
  >> 25    configure: error:
     26       -----------------------------------------------------------------------
     27         Specified MPI C compiler "cc" cannot be found under /opt/cray/pe/mpich/8.1.26/ofi/intel/2022.1
     28         PnetCDF requires a working MPI C compiler. Please specify the location
     29         of an MPI C compiler, either in the MPICC environment variable
     30         (not CC variable) or through --with-mpi configure flag. Abort.
     31       -----------------------------------------------------------------------

See build log for details:
  /p/work2/heinzell/spst-upd-spdev-20250415/cache/build_stage/spack-stage-parallel-netcdf-1.12.3-3issk2b4kp462j7u636t6vf774kppd5s/spack-build-out.txt
```

**Describe the solution you'd like**

1. Fix `parallel-netcdf` so that it still builds on Cray (because it is used by MPAS and mpas-jedi-env).
2. In parallel, prefer `parallelio ~pnetcdf` and enable the variant explicitly in `mpas-jedi-env` and others. I believe `ufs-weather-model-env` also wants this as the UFS is moving to MPAS in parts and won't have all the I/O capabilities in the native UFS IO layer (they may still be using the MPAS `parallelio`/`parallel-netcdf` I/O features for now)

**Additional context**

n/a

--- a/.github/workflows/ubuntu-ci-x86_64-gnu.yaml
+++ b/.github/workflows/ubuntu-ci-x86_64-gnu.yaml
@@ -51,7 +51,7 @@ jobs:
           # Set up spack-stack
           source ./setup.sh
 
-          declare -a TEMPLATES=("unified-dev" "skylab-dev" "cylc-dev")
+          declare -a TEMPLATES=("unified-dev" "skylab-dev" "cylc-dev" "neptune-dev")
           for TEMPLATE in "${TEMPLATES[@]}"; do
             if [[ "${TEMPLATE}" == *"unified-dev"* ]]; then
               export ENVNAME=ue-gcc-11.4.0-buildcache
@@ -59,6 +59,8 @@ jobs:
               export ENVNAME=se-gcc-11.4.0-buildcache
             elif [[ "${TEMPLATE}" == *"cylc-dev"* ]]; then
               export ENVNAME=ce-gcc-11.4.0-buildcache
+            elif [[ "${TEMPLATE}" == *"neptune-dev"* ]]; then
+              export ENVNAME=ne-gcc-11.4.0-buildcache
             fi
             echo "Creating environment ${ENVNAME} from template ${TEMPLATE}"
 

--- a/.github/workflows/ubuntu-ci-x86_64-gnu.yaml
+++ b/.github/workflows/ubuntu-ci-x86_64-gnu.yaml
@@ -99,6 +99,11 @@ jobs:
             # Add additional variants for MET packages, different from config/common/packages.yaml
             spack config add "packages:met:variants:+python +grib2 +graphics +lidar2nc +modis"
 
+            # Turn off variant xnrl for neptune-python-env because xnrl is not publicly available
+            if [[ "${TEMPLATE}" == *"neptune-dev"* ]]; then
+              sed -i 's/+xnrl/~xnrl/g' ${ENVDIR}/spack.yaml
+            fi
+
             # Concretize and check for duplicates
             spack concretize 2>&1 | tee log.concretize.gnu-11.4.0-buildcache
             ${SPACK_STACK_DIR}/util/show_duplicate_packages.py -d log.concretize.gnu-11.4.0-buildcache -i fms -i crtm -i crtm-fix -i esmf -i mapl

--- a/configs/common/packages.yaml
+++ b/configs/common/packages.yaml
@@ -211,7 +211,8 @@ packages:
   p4est:
     require: '@2.8.7'
   parallelio:
-    require: '@2.6.2 +pnetcdf'
+    require:
+    - '@2.6.2'
   parallel-netcdf:
     require: '@1.12.3'
   pflogger:

--- a/configs/templates/neptune-dev/spack.yaml
+++ b/configs/templates/neptune-dev/spack.yaml
@@ -22,3 +22,8 @@ spack:
       - neptune-python-env%intel
       # Same for LLVM compilers
       - neptune-python-env%clang
+  packages:
+    # Avoid duplicate builds of parallelio +pnetcdf and ~pnetcdf
+    parallelio:
+      require:
+      - '~pnetcdf'

--- a/configs/templates/skylab-dev/spack.yaml
+++ b/configs/templates/skylab-dev/spack.yaml
@@ -41,3 +41,8 @@ spack:
     exclude:
     # py-torch in ai-env doesn't build with Intel
     - ai-env%intel
+  packages:
+    # Avoid duplicate builds of parallelio +pnetcdf and ~pnetcdf
+    parallelio:
+      require:
+      - '+pnetcdf'

--- a/configs/templates/unified-dev/spack.yaml
+++ b/configs/templates/unified-dev/spack.yaml
@@ -59,3 +59,8 @@ spack:
     # Skip neptune-python-env with Intel Classic due to problems
     # with new versions of py-numpy, py-scipy, ...
     - neptune-python-env%intel
+  packages:
+    # Avoid duplicate builds of parallelio +pnetcdf and ~pnetcdf
+    parallelio:
+      require:
+      - '+pnetcdf'

--- a/spack-ext/repos/spack-stack/packages/base-env/package.py
+++ b/spack-ext/repos/spack-stack/packages/base-env/package.py
@@ -32,7 +32,6 @@ class BaseEnv(BundlePackage):
     depends_on("hdf5", type="run")
     depends_on("netcdf-c", type="run")
     depends_on("netcdf-fortran", type="run")
-    depends_on("parallel-netcdf", type="run")
     depends_on("parallelio", type="run")
     depends_on("nccmp", type="run")
 

--- a/spack-ext/repos/spack-stack/packages/jedi-mpas-env/package.py
+++ b/spack-ext/repos/spack-stack/packages/jedi-mpas-env/package.py
@@ -17,6 +17,8 @@ class JediMpasEnv(BundlePackage):
     version("1.0.0")
 
     depends_on("jedi-base-env", type="run")
+    depends_on("parallel-netcdf", type="run")
+    depends_on("parallelio +pnetcdf", type="run")
     depends_on("metis", type="run")
     depends_on("jasper", type="run")
 

